drivers/mbox/mbox_ti_secproxy.c

```diff
@@ -0,0 +1,263 @@
+ */
+
+#include <zephyr/spinlock.h>
+#include <stdio.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/irq.h>
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ti_secure_proxy);
+
+#define DT_DRV_COMPAT ti_secure_proxy
+
+#define DEV_CFG(dev)  ((const struct secproxy_mailbox_config *)((dev)->config))
+#define DEV_DATA(dev) ((struct secproxy_mailbox_data *)(dev)->data)
+
+#define DEV_TDATA(_dev) DEVICE_MMIO_NAMED_GET(_dev, target_data)
+#define DEV_RT(_dev)    DEVICE_MMIO_NAMED_GET(_dev, rt)
+#define DEV_SCFG(_dev)  DEVICE_MMIO_NAMED_GET(_dev, scfg)
+
+#define RT_THREAD_STATUS               0x0
+#define RT_THREAD_THRESHOLD            0x4
+#define RT_THREAD_STATUS_ERROR_SHIFT   31
+#define RT_THREAD_STATUS_ERROR_MASK    BIT(31)
+#define RT_THREAD_STATUS_CUR_CNT_SHIFT 0
+#define RT_THREAD_STATUS_CUR_CNT_MASK  GENMASK(7, 0)
+
+#define SCFG_THREAD_CTRL           0x1000
+#define SCFG_THREAD_CTRL_DIR_SHIFT 31
+#define SCFG_THREAD_CTRL_DIR_MASK  BIT(31)
+
+#define SEC_PROXY_THREAD(base, x) ((base) + (0x1000 * (x)))
+#define THREAD_IS_RX              1
+#define THREAD_IS_TX              0
+
+#define SECPROXY_MAILBOX_NUM_MSGS 5
+#define MAILBOX_MAX_CHANNELS      32
+#define MAILBOX_MBOX_SIZE         60
+
+struct secproxy_thread {
+	mem_addr_t target_data;
+	mem_addr_t rt;
+	mem_addr_t scfg;
+	char rx_data[MAILBOX_MBOX_SIZE * 4];
+};
+
+struct secproxy_mailbox_data {
+	mbox_callback_t cb[MAILBOX_MAX_CHANNELS];
+	void *user_data[MAILBOX_MAX_CHANNELS];
+	bool channel_enable[MAILBOX_MAX_CHANNELS];
+
+	DEVICE_MMIO_NAMED_RAM(target_data);
+	DEVICE_MMIO_NAMED_RAM(rt);
+	DEVICE_MMIO_NAMED_RAM(scfg);
+	struct k_spinlock lock;
+};
+
+struct secproxy_mailbox_config {
+	DEVICE_MMIO_NAMED_ROM(target_data);
+	DEVICE_MMIO_NAMED_ROM(rt);
+	DEVICE_MMIO_NAMED_ROM(scfg);
+	uint32_t irq;
+};
+
+static inline int secproxy_verify_thread(struct secproxy_thread *spt, uint8_t dir)
+{
+	/* Check for any errors already available */
+	if (sys_read32(spt->rt + RT_THREAD_STATUS) & RT_THREAD_STATUS_ERROR_MASK) {
+		LOG_ERR("Thread is corrupted, cannot send data.\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void secproxy_mailbox_isr(const struct device *dev)
+{
+	struct secproxy_mailbox_data *data = DEV_DATA(dev);
+	struct secproxy_thread spt;
+
+	for (int i_channel = 0; i_channel < MAILBOX_MAX_CHANNELS; i_channel++) {
+		if (!data->channel_enable[i_channel]) {
+			continue;
+		}
+
+		spt.target_data = SEC_PROXY_THREAD(DEV_TDATA(dev), i_channel);
+		spt.rt = SEC_PROXY_THREAD(DEV_RT(dev), i_channel);
+		spt.scfg = SEC_PROXY_THREAD(DEV_SCFG(dev), i_channel);
+
+		if (secproxy_verify_thread(&spt, THREAD_IS_RX)) {
+			LOG_ERR("Thread %d is in error state\n", i_channel);
+			continue;
+		}
+
+		/* check if there are messages */
+		if (!(sys_read32(spt.rt) & 0x7F)) {
+			continue;
+		}
+
+		for (int i = 4; i <= 0x3C; i += 4) {
+			*((uint32_t *)(spt.rx_data + i - 4)) = sys_read32(spt.target_data + i);
```

Comment:
This line is un-readable. Maybe split this into a couple lines, one to read data into a local var, and another to store into buffer. The compiler will eliminate the temp so no harm in making this readable, remember code is for humans :)